docs/proposals/06-immutable-etcd-backups.md

```diff
@@ -0,0 +1,292 @@
+- **Operator Intentions are Respected:** The ETCD cluster remains in its hibernated state, as per the operator's intention.
+- **Storage Costs are Controlled:** Avoids unnecessary extension of immutability for all snapshots, preventing increased storage costs.
+
+### New Compaction Controller Flag
+
+**Flag:** `--hibernation-snapshot-interval`
+
+- **Type:** Duration
+- **Default:** `24h`
+- **Description:** This flag sets the period after which a compaction job should be triggered for a hibernated ETCD cluster, based on the time since the last renewal of the full snapshot lease. If the time since `fullLease.Spec.RenewTime.Time` exceeds the duration specified by this flag, and `etcd.spec.replicas` is `0` (indicating hibernation), the compaction job will automatically trigger to create a new snapshot. This approach ensures that backups remain within the immutability period and are safeguarded against becoming mutable.
+
+### Garbage Collection during Compaction for Hibernated ETCD Clusters
+
+If an ETCD cluster is hibernated for a long duration, there is a chance that backup storage will accumulate a large number of snapshots. Since the ETCD cluster is not running, the `etcd-backup-restore` component is also not running to perform garbage collection of the snapshots.
+
+To address this, the compaction job should be enhanced to handle garbage collection during hibernation. This ensures that old snapshots are cleaned up appropriately, considering the immutability constraints.
+
+### Excluding Snapshots Under Specific Circumstances
+
+Given that immutable backups cannot be deleted until the immutability period expires, there are scenarios, such as corrupted snapshots or other anomalies, where certain snapshots must be skipped during the restoration process. To facilitate this:
+
+- **Custom Metadata Tags:** Utilize custom metadata to mark specific objects (snapshots) that should be bypassed. To exclude a snapshot from the restoration process, attach custom metadata to it with the key `x-etcd-snapshot-exclude` and value `true`. This method is officially supported, as demonstrated in the [etcd-backup-restore PR](https://github.com/gardener/etcd-backup-restore/pull/776).
+
+## Compatibility
+
+The proposed changes are fully compatible with existing ETCD clusters and backup processes. Operators are responsible for creating or updating the immutability settings on the backup storage buckets, but no changes are required to the ETCD clusters themselves.
+
+- **Backward Compatibility:** Existing clusters without immutable buckets will continue to function without change.
+- **Forward Compatibility:** Clusters can opt-in to use immutable backups by configuring the bucket accordingly.
+
+## Implementation Steps
+
+1. **Enhance the Compaction Job:**
+     - Modify the compaction job in `etcd-backup-restore` to manage backup processes for hibernated clusters, including snapshot creation and garbage collection, while considering the new immutability constraints.
```

Comment:
> Modify the compaction job in `etcd-backup-restore` to manage backup processes for hibernated clusters

snapshot compaction job won't distinguish b/w running cluster or hibernated cluster, so this statement doesn't make sense to me. All we need to handle is to how to trigger the compaction job for hibernated clusters but nothing changes for compact sub-command.

> including snapshot creation and garbage collection, while considering the new immutability constraints.

didn't get it, creation/upload of snapshot by compaction will remain unchange. Moreover, how's garbage collection coming into the picture of compaction job, I didn't get this part as well.

Reply:
This is outdated now, resolving.